[#49493] Add a `WorkPackageRole` type

aaron-contreras opened the pull request:
## Milestones
1. A new Role type will become available (`WorkPackageRole`) in order to capture/assign roles to users and groups appropriately for Work-Package-specific permissions in order to grant access to OpenProject on a Work-Package-specific basis.

2. Ability to specify/determine which role types permissions are assignable to. This will help simplify the form implementation for Role Creation.

3. Seed the three default Work Package roles and corresponding permissions that will come pre-packaged with OpenProject
  - Editor
  - Commenter
  - Viewer 

## Notes
Work Package: https://community.openproject.org/work_packages/49493

aaron-contreras committed:
Create `WorkPackageRole` model

* Executes:
  `./bin/rails g model WorkPackageRole --skip-migrations --parent Role`

* Tailors Factory and Model Spec

aaron-contreras committed:
Refactor: Split `RoleSeeder` into `Project` and `Global` Role seeders

In order to more easily accomodate adding a seeder for
 `WorkPackageRole`s, splitting the catch-all `RoleSeeder` into a
seeder per type would help facilitate and simplify how we grok through
role seeding.

The translation file structure was also modified to more easily group
the different role types that are to be created whenever we start
adding more role types to be seeded.

The structure is now:
```yaml
<<role_type>>s:
  # role_data
```

aaron-contreras commented on app/seeders/basic_data/base_role_seeder.rb:
```diff
@@ -74,7 +72,9 @@ def permissions(value)
-        Roles::CreateContract.new(Role.new, nil).assignable_permissions.map(&:name)
+        Roles::CreateContract.new(self.class.model_class.new, nil)
+                             .assignable_permissions
+                             .map(&:name)
```
I don't know if this is a bug or intended but this would always assign all `assignable_permissions` for Project Roles even if a `GlobalRole` is being seeded. @cbliard

cbliard replied:
That's not what I am observing. I tried it and it returned the expected global permissions.

We can have a look together. Ping me when you're available.

aaron-contreras commented on spec/seeders/basic_data/project_role_seeder_spec.rb:
```diff
@@ -103,42 +105,15 @@
         roles:
-        - reference: :role_project_admin
-          name: Project admin
-          global: true
```
@cbliard Following up on the `assignable_permissions` comment. This was intended to be a global role, but based on the spec it was expected to be assigned all `assignable_permissions` for the `Role` type.

Now that I've split these specs into `project_role_seeder_spec` and `global_role_seeder_spec` I've added an expectation on the `global_role_seeder_spec` that the assigned roles to the created role, only gets the `assignable_permissions` for `GlobalRole` types.

cbliard replied:
I think I did a mistake originally here. As it is the "Project admin" role, it should not have been `global: true`.
Thanks for having fixed it!

cbliard commented on app/seeders/common.yml:
```diff
@@ -240,111 +240,115 @@ open_color_palette:
+  global:
+    - reference: :default_role_project_creator_and_staff_manager
+      t_name: Staff and projects manager
+      global: true
```
`global: true` is not useful anymore.

cbliard commented on spec/support/seeders.rb:
```diff
@@ -48,7 +54,7 @@ def add_needed_seeders_for(seeder_class, needed_acc)
   shared_let(:needed_seeders_keys) do
     # warning: there is one limitation: RoleSeeder will only pick 'roles' key,
     # and not 'modules_permissions' key.
-    needed_seeders.map { _1.try(:seed_data_model_key) }.compact
```
I do not like introducing exceptions. It can then evolve to accidental complexity. The idea was to keep a 1-1 relation between the key name in the seed files and the class name of the record. This way I do not have to think too much to lookup the right key given a model name.

Instead of having `roles.project`, `roles.global`, and `roles.work_package` , why not use `project_roles`, `global_roles`, and `work_package_roles` directly? This way the role seeders would work as any other seeder.

What do you think about it?

cbliard commented on spec/seeders/basic_data/global_role_seeder_spec.rb:
```diff
@@ -0,0 +1,87 @@
+      SEEDING_DATA_YAML
+    end
+
+    it 'gives all assignable permissions to the role' do
+      expect(GlobalRole.find_by(name: 'Project admin').permissions)
+        .to match_array(Roles::CreateContract.new(GlobalRole.new, nil).assignable_permissions.map { _1.name.to_sym })
+    end
```
I think that's ok. When I run it in a console, it returns `[:add_project, :create_backup, :create_user, :manage_user, :manage_placeholder_user]`

Maybe the test can be enhanced to check that the permissions include `:add_project` and `:create_user`, and do not include `view_project` and `view_work_packages`. Just to be sure that they are global permissions and not project permissions.

(As a side note, I'm always suspicious when we assert equality using business code in the comparison: in this case, what if a bug is introduced in `Roles::CreateContract.new(GlobalRole.new, nil).assignable_permissions`? I suppose we have to trust that a unit test in `Roles::CreateContract` exists for when a `GlobalRole` is given. In that case, hardening the test would be useless. And after checking, it looks like `spec/contracts/roles/shared_contract_examples.rb` is covering it, so no need to add the "test enhancements" I suggested earlier: that would only duplicate code)

aaron-contreras replied:
Discussed in a call

cbliard commented on app/seeders/basic_data/base_role_seeder.rb:
```diff
@@ -43,8 +41,8 @@ def model_attributes(role_data)
-    def type(global_value)
-      true?(global_value) ? 'GlobalRole' : 'Role'
+    def type
+      self.class.model_class.to_s
```
`self.class` is not needed. `model_class.to_s` is enough.

cbliard commented on app/seeders/basic_data/base_role_seeder.rb:
```diff
@@ -74,7 +72,9 @@ def permissions(value)
         value
       when :all_assignable_permissions
-        Roles::CreateContract.new(Role.new, nil).assignable_permissions.map(&:name)
+        Roles::CreateContract.new(self.class.model_class.new, nil)
```
`self.class` not needed here.

aaron-contreras committed:
Create the `WorkPackageRoleSeeder` class

Creates the `WorkPackageRoleSeeder` class in charge of seeding the
default roles of the `WorkPackageRole` type.

Seeder specs added to assert the expected behavior.

aaron-contreras committed:
Seed default `WorkPackageRole`s

Seeds the 3 default `WorkPackageRole`s with the set of permissions
present on `config/initializers/permissions.rb` that apply.

aaron-contreras committed:
Add `costs` module permissions to `WorkPackageRole`s

aaron-contreras committed:
Add `storages` module permissions to `WorkPackageRole`s

aaron-contreras committed:
Add `github_integration` module permissions to `WorkPackageRole`s

klaustopher reviewed (commented):
Looked through everything. Very thorough!

aaron-contreras committed:
Refactor `permissions_spec.rb`

* Replace usage of `match_array` with `contain_exactly`.

* Use `subject` construct instead of `let`.

* Reword contexts for verbosity and clarity.

aaron-contreras committed:
Refactor `access_control_spec.rb`

* Clarifies intent and behavior of the module.

* Addresses some Rubocop warnings.

aaron-contreras committed:
Refactor: Rename `unstash_instance_variables`

Reads a bit better to me :)

aaron-contreras committed:
Add `:permissible_on` parameter to `Permission` constructor

Allows defining what role types a permission is assignable to. This
provides a more flexible infrastructure than `global: true` or
`global: false` now that we're moving on to have more than two role
types.

aaron-contreras committed:
Chore: Fix Rubocop anonymous argument forwarding cop

* Fixes:
```
RuboCop: Use anonymous keyword arguments forwarding (`**`).
[Style/ArgumentsForwarding]
```

aaron-contreras committed:
Refactor: Set `permissible_on` definitions to `access_control_spec.rb`

aaron-contreras committed:
Set `permissible_on` definitions to `nil` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Create `copy_work_packages` permission

Throughout some specs, we already set up users with a
`copy_work_packages` permission. This led me to believe that at some
point, there was the intent to split `move_work_packages` up to have
a more fine-grained control between the two.

Now, we would need the `copy_work_packages` permission for Work Package
sharing, instead of the `move_work_packages` permission so we don't
have a permission escalation concern and can manually assert against
`copy_work_packages`.

* Adds permission definition to initializer file

* Creates migration to add the permission to those roles which already
  have the broader `move_work_packages`

* Adds `copy_work_packages` to the `work_package_roles` seed data to
  populate the new default work package roles.

* Add translation for permission.

aaron-contreras committed:
Set `permissible_on` definitions to `work_package_tracking` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Set `permissible_on` definitions to `news` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Set `permissible_on` definitions to `wiki` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Set `permissible_on` definitions to `repository` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Set `permissible_on` definitions to `forums` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Set `permissible_on` definitions to `activity` module in core initializer

* Includes `work_package`, `project` and `global` permissible_on
  definitions.

aaron-contreras committed:
Set `permissible_on` definitions on `backlogs` module

aaron-contreras committed:
Set `permissible_on` definitions on `bim` module

aaron-contreras committed:
Chore: Fix Rubocop indentation cop on `boards/engine.rb`

* Fixes:
```
RuboCop: Use 2 (not 4) spaces for indentation.
[Layout/IndentationWidth]
```

aaron-contreras committed:
Set `permissible_on` definitions on `boards` module

aaron-contreras committed:
Set `permissible_on` definitions on `budgets` module

aaron-contreras committed:
Chore: Fix Rubocop indentation cop on `calendars/engine.rb`

* Fixes:
```
RuboCop: Use 2 (not 4) spaces for indentation.
[Layout/IndentationWidth]
```

aaron-contreras committed:
Set `permissible_on` definitions on `calendars` module

aaron-contreras committed:
Set `permissible_on` definitions on `costs` module

klaustopher commented on lib/open_project/access_control/permission.rb:
```diff
@@ -42,6 +42,7 @@ def initialize(name,
                      hash,
                      public: false,
                      require: nil,
+                     permissible_on: [],
```
I'd opt to remove the default value here and make it mandatory to add this to a  permission definiton

aaron-contreras replied:
Ah, yes! I was thinking about adding a warning somewhere in case someone forgets about defining it but the error message raised by removing the default is much better. :)

aaron-contreras committed:
Set `permissible_on` definitions on `dashboards` module

aaron-contreras committed:
Chore: Fix Rubocop file path cop on `dashboards/engine.rb`

Fixes:
```
RuboCop: Prefer `Rails.root.join('path/to')`. [Rails/FilePath]
```

aaron-contreras committed:
Set `permissible_on` definitions on `documents` module

aaron-contreras committed:
Chore: Fix Rubocop redundant current directory cop in `github_integration/engine.rb`

Fixes:
```
Style/RedundantCurrentDirectoryInPath:
Remove the redundant current directory path.
```

aaron-contreras committed:
Set `permissible_on` definitions on `github_integration` module

aaron-contreras committed:
Chore: Fix Rubocop warning cops in `meeting/engine.rb`

Fixes:
```
Layout/IndentationWidth: Use 2 (not 4) spaces for indentation.
```

aaron-contreras committed:
Set `permissible_on` definitions on `meeting` module

aaron-contreras committed:
Chore: Add copyright to `overviews/engine.rb`

aaron-contreras committed:
Set `permissible_on` definitions on `overviews` module

aaron-contreras committed:
Set `permissible_on` definitions on `reporting` module

aaron-contreras committed:
Set `permissible_on` definitions on `storages` module

aaron-contreras committed:
Chore: Fix Rubocop warning cops in `team_planner/engine.rb`

Fixes:
```
Layout/IndentationWidth: Use 2 (not 4) spaces for indentation.
```

aaron-contreras committed:
Chore: Update Copyright in `team_planner/engine.rb`

aaron-contreras committed:
Set `permissible_on` definitions on `team_planner` module

aaron-contreras committed:
Refactor: Update implementation of `mock_global_permissions` helper

aaron-contreras committed:
Refactor: Rename variables in `actions/scopes/default_spec.rb`

I had to really read through it to understand it the first time around.

Hope this helps someone else (or myself) down the line when it's
re-visited.

aaron-contreras committed:
Add `.project_permissions` method to `OpenProject::AccessControl`

* Allows querying specifically for project permissions.

aaron-contreras committed:
Refactor: Enhance `doc` format of RSpec output

This reads better in documentation format output.

aaron-contreras committed:
Add `.project_permissions` method to `OpenProject::AccessControl`

* Allows querying specifically for project permissions.

* Adds spec to ensure permissions that are permissible on multiple
  role types show up in their appropriate query.

aaron-contreras committed:
Improvement: Access `.permissions` instead of `@mapped_permissions`

I can't see a use-case in which we'd want to assign, consider or
take into account the subset of permissions which are not enabled
during boot or reload.

This commit standardizes access/manipulation to those permissions which
are marked as responding to `#enabled? => truthy`.
subset of non-enabled permissions

aaron-contreras committed:
Add `#assignable_work_package_permissions` to Roles Base Contract

* Refactors `#assignable_member_permissions` to use the new
  `project_permissions`.

* Adds `#assignable_work_package_permissions`.

aaron-contreras committed:
Update overview module `roles/base_contract_spec.rb`

Adds in a spec to test for `assignable_work_package_permissions`.

aaron-contreras committed:
Scope visibility to Non Work Package Roles in Settings

It's agreed that the settings panel should have no notion that
Work Package Roles even exist until we decide to make them editable
by users.

aaron-contreras committed:
Chore: Rename Example Group description and migrate to cuprite

klaustopher reviewed (approved):
Awesome work!
